README.md
# Kraken

<h1 align="center">
  <br>
  <img src="https://raw.githubusercontent.com/kraken-ng/Kraken/main/static/kraken-logo-background.jpg" alt="Kraken">
</h1>

<h4 align="center">Kraken, a modular multi-language webshell coded by @secu_x11.</h4>

<p align="center">
  <a href="https://github.com/kraken-ng/Kraken/wiki/Getting-Started#requirements">Requirements</a> •
  <a href="https://github.com/kraken-ng/Kraken/wiki/Support">Support</a> •
  <a href="https://github.com/kraken-ng/Kraken/wiki/Getting-Started#installation">Install</a> •
  <a href="https://github.com/kraken-ng/Kraken/wiki/Getting-Started#usage">Usage</a> •
  <a href="https://github.com/kraken-ng/Kraken/wiki/Getting-Started#advanced-usage">Advanced Usage</a> •
  <a href="https://github.com/kraken-ng/Kraken/wiki/Contribute">Contributing</a> •
  <a href="https://github.com/kraken-ng/Kraken/wiki/FAQ-&-Troubleshooting">FAQ</a> •
  <a href="https://github.com/kraken-ng/Kraken/wiki/Acknowledgments-&-References">Acknowledgments</a>
</p>

---

Kraken is a modular multi-language webshell focused on web post-exploitation and defense evasion. It supports three technologies (**PHP, JSP and ASPX**) and is core is developed in Python.

Kraken follows the principle of <b><ins>"avoiding command execution"</ins></b> by re-implementing it through the functionalities of the programming language in use. Kraken seeks to provide usability, scalability and improve the OPSEC of ongoing operations.

Although its main use is focused on offensive purposes (e.g. red teams, internal pentest), it is possible to use it by blue teams to evaluate existing defensive tools and configurations.

## Version

1.2.0 - [CHANGELOG Version](CHANGELOG.md)

## Documentation

All the information about the installation process, usage, internal design, contributions and much more is explained in the [Github Wiki]:

- **[I. Home]**
- **[II. Support]**
  - [Agents supported]
  - [Modules supported]
    - [Support on Linux]
    - [Support on Windows]
  - [Version testing]
- **[III. Getting Started]**
  - [Requirements]
  - [Installation]
    - [Installation on Linux]
    - [Installation on Windows]
    - [Installation on MacOS]
  - [Usage]
    - [Choosing the web technology]
    - [Choosing the environment]
    - [Deploying the environment]
    - [Modifying agent]
    - [Selecting a connection profile]
    - [Connection from client]
  - [Advanced Usage]
    - [Standard Mode]
      - [Using Java Agent]
      - [Using NET Agent]
    - [Command and control Mode]
- **[IV. Contribute]**
  - [Module contributions]
    - [Introduction]
    - [PHP Module Creation]
      - [PHP Template Structure]
      - [PHP Module Development]
      - [PHP Module Versioning]
      - [PHP Module Registration]
    - [Java Module Creation]
      - [Java Template Structure]
      - [Java Module Development]
      - [Java Module Versioning]
      - [Java Module Registration]
    - [NET Module Creation]
      - [NET Template Structure]
      - [NET Module Development]
      - [NET Module Versioning]
      - [NET Module Registration]
  - [Dispatcher contributions]
  - [Formater contributions]
  - [Environment contributions]
  - [Util contributions]
- **[V. Internals]**
- **[VI. FAQ & Troubleshooting]**
  - [Known bugs]
- **[VII. Acknowledgments & References]**
  - [Acknowledgments]
  - [References]




[Github Wiki]: https://github.com/kraken-ng/Kraken/wiki
[I. Home]: https://github.com/kraken-ng/Kraken/wiki
[II. Support]: https://github.com/kraken-ng/Kraken/wiki/Support
[Agents supported]: https://github.com/kraken-ng/Kraken/wiki/Support#agents
[Modules supported]: https://github.com/kraken-ng/Kraken/wiki/Support#modules
[Support on Linux]: https://github.com/kraken-ng/Kraken/wiki/Support#support-on-linux
[Support on Windows]: https://github.com/kraken-ng/Kraken/wiki/Support#support-on-windows
[Version testing]: https://github.com/kraken-ng/Kraken/wiki/Support#version-testing
[III. Getting Started]: https://github.com/kraken-ng/Kraken/wiki/Getting-Started
[Requirements]: https://github.com/kraken-ng/Kraken/wiki/Getting-Started#requirements
[Installation]: https://github.com/kraken-ng/Kraken/wiki/Getting-Started#installation
[Installation on Linux]: https://github.com/kraken-ng/Kraken/wiki/Getting-Started#installation-on-linux
[Installation on Windows]: https://github.com/kraken-ng/Kraken/wiki/Getting-Started#installation-on-windows
[Installation on MacOS]: https://github.com/kraken-ng/Kraken/wiki/Getting-Started#installation-on-macos
[Usage]: https://github.com/kraken-ng/Kraken/wiki/Getting-Started#usage
[Choosing the web technology]: https://github.com/kraken-ng/Kraken/wiki/Getting-Started#choosing-the-web-technology
[Choosing the environment]: https://github.com/kraken-ng/Kraken/wiki/Getting-Started#choosing-the-environment
[Deploying the environment]: https://github.com/kraken-ng/Kraken/wiki/Getting-Started#deploying-the-environment
[Modifying agent]: https://github.com/kraken-ng/Kraken/wiki/Getting-Started#modifying-agent
[Selecting a connection profile]: https://github.com/kraken-ng/Kraken/wiki/Getting-Started#selecting-a-connection-profile
[Connection from client]: https://github.com/kraken-ng/Kraken/wiki/Getting-Started#connection-from-client
[Advanced Usage]: https://github.com/kraken-ng/Kraken/wiki/Getting-Started#advanced-usage
[Standard Mode]: https://github.com/kraken-ng/Kraken/wiki/Getting-Started#standard-mode
[Using Java Agent]: https://github.com/kraken-ng/Kraken/wiki/Getting-Started#using-java-agent
[Using NET Agent]: https://github.com/kraken-ng/Kraken/wiki/Getting-Started#using-net-agent
[Command and control Mode]: https://github.com/kraken-ng/Kraken/wiki/Getting-Started#command-and-control-mode
[IV. Contribute]: https://github.com/kraken-ng/Kraken/wiki/Contribute
[Module contributions]: https://github.com/kraken-ng/Kraken/wiki/Contribute#modules
[Introduction]: https://github.com/kraken-ng/Kraken/wiki/Contribute#introduction
[PHP Module Creation]: https://github.com/kraken-ng/Kraken/wiki/Contribute#php-module-creation
[PHP Template Structure]: https://github.com/kraken-ng/Kraken/wiki/Contribute#php-template-structure
[PHP Module Development]: https://github.com/kraken-ng/Kraken/wiki/Contribute#php-module-development
[PHP Module Versioning]: https://github.com/kraken-ng/Kraken/wiki/Contribute#php-module-versioning
[PHP Module Registration]: https://github.com/kraken-ng/Kraken/wiki/Contribute#php-module-registration
[Java Module Creation]: https://github.com/kraken-ng/Kraken/wiki/Contribute#java-module-creation
[Java Template Structure]: https://github.com/kraken-ng/Kraken/wiki/Contribute#java-template-structure
[Java Module Development]: https://github.com/kraken-ng/Kraken/wiki/Contribute#java-module-development
[Java Module Versioning]: https://github.com/kraken-ng/Kraken/wiki/Contribute#java-module-versioning
[Java Module Registration]: https://github.com/kraken-ng/Kraken/wiki/Contribute#java-module-registration
[NET Module Creation]: https://github.com/kraken-ng/Kraken/wiki/Contribute#net-module-creation
[NET Template Structure]: https://github.com/kraken-ng/Kraken/wiki/Contribute#net-template-structure
[NET Module Development]: https://github.com/kraken-ng/Kraken/wiki/Contribute#net-module-development
[NET Module Versioning]: https://github.com/kraken-ng/Kraken/wiki/Contribute#net-module-versioning
[NET Module Registration]: https://github.com/kraken-ng/Kraken/wiki/Contribute#net-module-registration
[Dispatcher contributions]: https://github.com/kraken-ng/Kraken/wiki/Contribute#dispatchers
[Formater contributions]: https://github.com/kraken-ng/Kraken/wiki/Contribute#formaters
[Environment contributions]: https://github.com/kraken-ng/Kraken/wiki/Contribute#environments
[Util contributions]: https://github.com/kraken-ng/Kraken/wiki/Contribute#utils
[V. Internals]: https://github.com/kraken-ng/Kraken/wiki/Internals
[VI. FAQ & Troubleshooting]: https://github.com/kraken-ng/Kraken/wiki/FAQ-&-Troubleshooting
[Known bugs]: https://github.com/kraken-ng/Kraken/wiki/FAQ-&-Troubleshooting#known-bugs
[VII. Acknowledgments & References]: https://github.com/kraken-ng/Kraken/wiki/Acknowledgments-&-References
[Acknowledgments]: https://github.com/kraken-ng/Kraken/wiki/Acknowledgments-&-References#acknowledgments
[References]: https://github.com/kraken-ng/Kraken/wiki/Acknowledgments-&-References#references


CHANGELOG.md
# Change Log

## [v1.2.0] - 2023-06-29

- New agents added
  - PHP Agent using "create_function" as executor
  - PHP Agent using "include/require" as executor
  - NET Agent using "Assembly.Load" as executor
  - NET Agent using "System.Reflection.Emit" as executor
- New compilers added to be used with different executors:
  - "Csc": this compiler allows you to compile each module you want to use at runtime and send it to an agent with an in-memory NET Assemblies loader (such as a Reflective NET Loader).
  - "Precompiled": this compiler allows you to use pre-compiled versions of .NET modules in .exe or .dll (NET Assemblies) form. You can use this compiler together with agent with an in-memory NET Assemblies loader.
- New utils:
  - "Precompiler": utility that allows to precompile Kraken modules. These can be used in executors that use dynamic loading of binaries into memory.
- New NET Assemblies:
  - New NET Assemblies focused on Privilege Escalation abusing DCOM services have been added: PrinterNotifyPotato and McpManagementPotato.
  - A modified version of EfsPotato, an Elevation of Privilege exploit abusing MS-EFSR, has been added.
- New modules:
  - "enum_antivirus": module to enumerate registered antivirus (via WMI) in Windows systems. Readapted from [Seatbelt command](https://github.com/GhostPack/Seatbelt/blob/master/Seatbelt/Commands/Windows/AntiVirusCommand.cs)
- By introducing new agents and compilers, small changes have been made in the Core and in the different submodules (envs, modules, etc).
- Multiple Bugs Fixed
  - Multiple minor bugs affecting some net_assemblies have been fixed. Some net assemblies have been adapted so that all of them work correctly with the execute_assembly module.
  - Fixed some minor bugs in PHP agents (st and c2). They now provide more insight into certain bugs.
  - Fixed a minor bug in PHP execute module. This change prevents a warning from occurring when using utf8_encode() on Windows systems whose PHP version indicates that this function is deprecated.

## [v1.1.1] - 2023-03-24

- Multiple bugs fixed:
  - Adapted Kraken ASP NET agent to work in NET Framework 2.0.
  - Adapted Kraken's modules to work in NET Framework 2.0.
  - Minimal fixes in NET modules.
  - Minimal fixes in some Kraken's utils.
  - Extending versioning in some NET Assemblies.

## [v1.1.0] - 2023-03-22

- Multiple bugs fixed:
  - "cp" (Java)
    - A missing try catch block has been added to control possible exceptions thrown during path normalization.
  - "dup_token" (NET)
    - The handle to the target process has been changed by using the OpenProcess function (which gives more control and allows specifying a flag to interact with PPL processes). Thanks to [Pablo Martinez](https://twitter.com/xassiz) for the notice and support.
    - We have also eliminated a block for checking the identity of the Current Thread that, on certain occasions, produced an unexpected error.
  - "execute" (JAVA)
    - A missing try catch block has been added to control possible exceptions thrown during path normalization.
  - "execute_with_token" (NET)
    - The CreateProcessWithToken function has been changed to CreateProcessAsUser. In previous versions, this module did not work correctly when using tokens from non-NT Authority System users. Thanks to the fix added by [Kurosh](https://twitter.com/_Kudaes_) in Hanshell I was able to port it to Kraken.
    - The same identity check block in the "dup_token" module has also been removed.
  - "reg_dump_trans" (NET), previously as secretsdump. (cc: @xassiz)
    - The name has been changed because the functionality that this module performed could be extended and, therefore, its functionality changed.
    - Extraction of any registry key is now supported.
    - Some missing functionalities have been added for the closing of the elements used by the module (transacted file rollback, closing of handlers and mapped view file).
    - The check based on administrative permissions has been changed to one based on SeBackup privilege. Since not all registry key extractions require administrative privileges, however, SeBackup is needed.
  - Fixed a bug in the standard Java agent that prevented correct operation in certain environments (Liferay).
  - Fixed a bug in the processing of the HTTP response returned by the agent.
- New modules:
  - "hotfixes": module to show hotfixes installed on current computer (via WMI). Thanks to [Etnum](https://twitter.com/the_etnum)
- New environments:
  - Different versions of Liferay environments have been added where the correct operation of Kraken Java agents has been validated..

## [v1.0.1] - 2023-03-13

- Added Kraken Wiki (with detailed information about the installation process and usage, as well as contributions in Kraken modules).
- New modules:
  - "mv": module to move files or directories.
- Mutiple bugs fixed:
  - Fixed some path normalization issues in all Java modules as well as in the template. 

## [v1.0.0] - 2023-02-25

- Multiple bugs fixed:
  - Adapted Kraken's modules to work in "1.6 >= Java <= 17" versions.
  - Fixed some path normalization issues in all PHP modules as well as in the template.
- New core features:
  - The "references" parameter has been added to the Kraken message passing structure. This parameter, which only works in NET, allows the specific references of each module to be added in the compilation process (csc). This avoids referencing unnecessary DLLs during this process. It also avoids a compilation error when there is no DLL in the system, as well as it allows the possibility of "linking" DLLs that the operator uploads directly to the target machine.
- New modules:
  - "sc": module to control services in NET implants using Service Controller functions.
- Updates in modules:
  - Java support has been added to the "id" module.
  - The "powerpick" module has been refactored and some existing bugs have been fixed.
  - PHP support has been added to the "tcpconnect" module.


lib/config.py
# Kraken modes
MODE_STANDARD                    = "st"
MODE_C2                          = "c2"

# Action codes
ACTION_STATUS                    = "0"
ACTION_INVOKE                    = "1"
ACTION_LOAD                      = "2"
ACTION_UNLOAD                    = "3"
ACTION_CLEAN                     = "4"

# Response codes
SUCCESS_RESPONSE_CODE            = "0"
FAILED_RESPONSE_CODE             = "1"

# Core Commands
COMMAND_EXIT                     = "exit"
COMMAND_HELP                     = "help"
COMMAND_INFO                     = "info"
COMMAND_CHANGE_DIRECTORY         = "cd"
COMMAND_REV2SELF                 = "rev2self"
COMMAND_CLOSE_TOKEN              = "close_token"
COMMAND_RECOMPILE                = "recompile"
COMMAND_LOGGING                  = "logging"

# Exclusive C2 Commands
C2_COMMAND_LIST_MODULES          = "list_modules"
C2_COMMAND_LOAD_MODULE           = "load_module"
C2_COMMAND_UNLOAD_MODULE         = "unload_module"
C2_COMMAND_REFRESH_MODULES       = "refresh_modules"
C2_COMMAND_CLEAN_MODULES         = "clean_modules"

# Configuration values for the HTTP client
HTTP_RESPONSE_SUCC               = 200
HTTP_MAX_HTTP_PARAM_LEN          = 8192
HTTP_MAX_HTTP_HEADER_LEN         = 8192
SUPPORTED_HTTP_METHODS           = ["GET", "POST"]
SUPPORTED_HTTP_ENCAPSULATE_TYPES = ["HEADER", "COOKIE", "FIELD"]
HTTP_AGENT_INVALID_METHOD        = 400
HTTP_AGENT_INVALID_AUTH          = 500

# Separators availables
FIELD_SEPARATOR                  = ","
VALUE_SEPARATOR                  = "="
CUSTOM_SEPARATOR                 = "|"

# Configuration values for paths
CLIENTS_PATH                     = "clients"
HISTORY_PATH                     = "logs"
MODULES_PATH                     = "modules"
TEMPORAL_PATH                    = "temp"
COMPILERS_PATH                   = "compilers"
DISPATCHERS_PATH                 = "dispatchers"
FORMATERS_PATH                   = "formaters"
AGENTS_PATH                      = "agents"
ENVS_PATH                        = "envs"
UTILS_PATH                       = "utils"

# Other configuration values
DEFAULT_EMPTY_EXECUTOR           = "-"
DEFAULT_TOKEN_VALUE              = "0"
RE_IMPERSONATE_TOKEN             = r"Impersonated\suser:\s'(.*?)'\swith\stoken:\s'(\d+)'"
DATETIME_FORMAT                  = "%Y/%m/%d %H:%M:%S"
MODULE_CLASS_NAME                = "Module"
ALL_MODULES                      = "all"
AVAILABLE_SO_AGENTS              = ["Linux", "Windows"]
AVAILABLE_TYPE_AGENTS            = ["php", "java", "cs"]
AVAILABLE_COMPILERS              = ["raw", "container", "precompiled", "csc"]

# Executors Configuration (and supported compilers)
EXECUTORS = {
    "php" : [
        {
            "name" : "eval",
            "id" : 0,
            "compilers" : ["raw"],
        },
        {
            "name" : "create_function",
            "id" : 3,
            "compilers" : ["raw"],
        },
        {
            "name" : "include",
            "id" : 4,
            "compilers" : ["raw"],
        },
        {
            "name" : "require",
            "id" : 5,
            "compilers" : ["raw"],
        },
    ],
    "java" : [
        {
            "name" : "ClassLoader",
            "id" : 1,
            "compilers" : ["container"],
        },
    ],
    "cs" : [
        {
            "name" : "CSharpCodeProvider",
            "id" : 2,
            "compilers" : ["raw"],
        },
        {
            "name" : "Assembly.Load",
            "id" : 6,
            "compilers" : ["precompiled", "csc"],
        },
        {
            "name" : "System.Reflection.Emmit",
            "id" : 7,
            "compilers" : ["precompiled", "csc"],
        },
    ]
}

# Containers Configuration
CONTAINERS = {
    "java" : [
        {
            "version" : "1.6",
            "image" : "openjdk:6"
        },
        {
            "version" : "1.7",
            "image" : "openjdk:7"
        },
        {
            "version" : "1.8",
            "image" : "openjdk:8"
        },
        {
            "version" : "9",
            "image" : "openjdk:9"
        },
        {
            "version" : "10",
            "image" : "openjdk:10"
        },
        {
            "version" : "11",
            "image" : "openjdk:11"
        },
        {
            "version" : "12",
            "image" : "openjdk:12"
        },
        {
            "version" : "13",
            "image" : "openjdk:13"
        },
        {
            "version" : "14",
            "image" : "openjdk:14"
        },
        {
            "version" : "15",
            "image" : "openjdk:15"
        },
        {
            "version" : "16",
            "image" : "openjdk:16"
        },
        {
            "version" : "17",
            "image" : "openjdk:17"
        }
    ],
    "php" : [],
    "cs" : []
}

# JSON Schema for connection profile
PROFILE_SCHEMA = {
    "$schema": "http://json-schema.org/draft-04/schema#",
    "type": "object",
    "properties": {
        "client": {
            "type": "object",
            "properties": {
                "url": {
                    "type": "string"
                },
                "method": {
                    "type": "string"
                },
                "headers": {
                    "type": "object"
                },
                "cookies": {
                    "type": "object"
                },
                "fields": {
                    "type": "object"
                },
                "message": {
                    "type": "object",
                    "properties": {
                        "secret": {
                            "type": "object",
                            "properties": {
                                "type": {
                                    "type": "string"
                                },
                                "key": {
                                    "type": "string"
                                },
                                "value": {
                                    "type": "string"
                                }
                            },
                            "required": [
                                "type",
                                "key",
                                "value"
                            ]
                        },
                        "data": {
                            "type": "object",
                            "properties": {
                                "type": {
                                    "type": "string"
                                },
                                "key": {
                                    "type": "string"
                                }
                            },
                            "required": [
                                "type",
                                "key"
                            ]
                        }
                    },
                    "required": [
                        "secret",
                        "data"
                    ]
                }
            },
            "required": [
                "url",
                "method",
                "headers",
                "cookies",
                "fields",
                "message"
            ]
        },
        "server": {
            "type": "object",
            "properties": {
                "type": {
                    "type": "string"
                },
                "key": {
                    "type": "string"
                }
            },
            "required": [
                "type",
                "key"
            ]
        }
    },
    "required": [
        "client",
        "server"
    ]
}

# Core Commands Configuration
CORE_COMMANDS = [
    {
        "name" : "exit",
        "description" : "Ends the session with the agent",
        "author" : "@secu_x11",
        "examples" : [
            "exit"
        ],
        "so" : [
            {
                "name" : "Linux",
                "agents" : ["php","java"]
            },
            {
                "name" : "Windows",
                "agents" : ["php","java","cs"]
            }
        ],
        "args" : []
    },
    {
        "name" : "help",
        "description" : "Displays the available commands or information of a command passed as an argument",
        "author" : "@secu_x11",
        "examples" : [
            "help",
            "help ls"
        ],
        "so" : [
            {
                "name" : "Linux",
                "agents" : ["php","java"]
            },
            {
                "name" : "Windows",
                "agents" : ["php","java","cs"]
            }
        ],
        "args" : [
            {
                "module": {
                    "help": "Module to show helper",
                    "nargs" : 1,
                    "type":  str
                }
            }
        ]
    },
    {
        "name" : "info",
        "description" : "Show Information about the Kraken's context (client and agent config)",
        "author" : "@secu_x11",
        "examples" : [
            "info"
        ],
        "so" : [
            {
                "name" : "Linux",
                "agents" : ["php","java"]
            },
            {
                "name" : "Windows",
                "agents" : ["php","java","cs"]
            }
        ],
        "args" : []
    },
    {
        "name" : "logging",
        "description" : "Turn on/off the extended logging",
        "author" : "@secu_x11",
        "examples" : [
            "logging"
        ],
        "so" : [
            {
                "name" : "Linux",
                "agents" : ["php","java"]
            },
            {
                "name" : "Windows",
                "agents" : ["php","java","cs"]
            }
        ],
        "args" : []
    },
    {
        "name" : "recompile",
        "description" : "Recompile one or multiple modules (use 'all' to recompile all modules)",
        "author" : "@secu_x11",
        "examples" : [
            "recompile ls",
            "recompile ls cat id ps",
            "recompile all"
        ],
        "so" : [
            {
                "name" : "Linux",
                "agents" : ["java"]
            },
            {
                "name" : "Windows",
                "agents" : ["java"]
            }
        ],
        "args" : [
            {
                "module": {
                    "help": "Module/s to recompile",
                    "nargs" : "*",
                    "type":  str
                }
            }
        ]
    },
    {
        "name" : "rev2self",
        "description" : f"Revert back to the original security context (sets the token to {DEFAULT_TOKEN_VALUE})",
        "author" : "@secu_x11",
        "examples" : [
            "rev2self"
        ],
        "so" : [
            {
                "name" : "Windows",
                "agents" : ["cs"]
            }
        ],
        "args" : []
    }
]

# C2 Command configuration
C2_COMMANDS = [
    {
        "name" : "list_modules",
        "description" : "List modules loaded in the agent",
        "author" : "@secu_x11",
        "examples" : [
            "list_modules"
        ],
        "so" : [
            {
                "name" : "Linux",
                "agents" : ["php"]
            },
            {
                "name" : "Windows",
                "agents" : ["php"]
            }
        ],
        "args" : []
    },
    {
        "name" : "load_module",
        "description" : "Load a new agent module/s (use 'all' to load all available modules)",
        "author" : "@secu_x11",
        "examples" : [
            "load_module ls",
            "load_module ls cd cat id",
            "load_module all"
        ],
        "so" : [
            {
                "name" : "Linux",
                "agents" : ["php"]
            },
            {
                "name" : "Windows",
                "agents" : ["php"]
            }
        ],
        "args" : [
            {
                "module": {
                    "help": "Module/s to load",
                    "nargs" : "*",
                    "type":  str
                }
            }
        ]
    },
    {
        "name" : "unload_module",
        "description" : "Unload an existing agent module/s (use 'all' to unload all loaded modules)",
        "author" : "@secu_x11",
        "examples" : [
            "unload_module ls",
            "unload_module ls cd cat id",
            "unload_module all"
        ],
        "so" : [
            {
                "name" : "Linux",
                "agents" : ["php"]
            },
            {
                "name" : "Windows",
                "agents" : ["php"]
            }
        ],
        "args" : [
            {
                "module": {
                    "help": "Module/s to unload",
                    "nargs" : "*",
                    "type":  str
                }
            }
        ]
    },
    {
        "name" : "refresh_modules",
        "description" : "Refresh module status from agent (update agent's memory in use)",
        "author" : "@secu_x11",
        "examples" : [
            "refresh_modules"
        ],
        "so" : [
            {
                "name" : "Linux",
                "agents" : ["php"]
            },
            {
                "name" : "Windows",
                "agents" : ["php"]
            }
        ],
        "args" : []
    },
    {
        "name" : "clean_modules",
        "description" : "Unload all agent modules",
        "author" : "@secu_x11",
        "examples" : [
            "clean_modules"
        ],
        "so" : [
            {
                "name" : "Linux",
                "agents" : ["php"]
            },
            {
                "name" : "Windows",
                "agents" : ["php"]
            }
        ],
        "args" : []
    }
]
